test/integration/iterator/prefix_test.go

```diff
@@ -31,3 +32,52 @@ func TestIteratorPrefix(t *testing.T) {
 
 	test.Execute(t)
 }
+
+func TestIteratorPrefix_DoesNotReturnSelf_Memory(t *testing.T) {
+	test := &integration.Test{
+		SupportedStoreTypes: []state.StoreType{
+			state.MemoryStoreType,
+		},
+		Actions: []action.Action{
+			action.Set([]byte("k"), []byte("v")),
+			action.Set([]byte("k1"), []byte("v1")),
+			&action.Iterate{
+				IterOptions: corekv.IterOptions{
+					Prefix: []byte("k"),
+				},
+				Expected: []action.KeyValue{
+					// `k` must not be yielded, as prefxes do not contain themselves
+					{Key: []byte("k1"), Value: []byte("v1")},
+				},
+			},
+		},
+	}
+
+	test.Execute(t)
+}
+
+// This test documents unwanted behaviour, it is tracked by:
+// https://github.com/sourcenetwork/corekv/issues/27
+func TestIteratorPrefix_DoesNotReturnSelf_Badger(t *testing.T) {
+	test := &integration.Test{
+		SupportedStoreTypes: []state.StoreType{
+			state.BadgerStoreType,
+		},
+		Actions: []action.Action{
+			action.Set([]byte("k"), []byte("v")),
+			action.Set([]byte("k1"), []byte("v1")),
+			&action.Iterate{
+				IterOptions: corekv.IterOptions{
+					Prefix: []byte("k"),
+				},
+				Expected: []action.KeyValue{
+					// `k` should not be yielded, but it is.
+					{Key: []byte("k"), Value: []byte("v")},
```

Comment:
question: Why is this unwanted behaviour? If it didn't return a key matching exactly the prefix, I would think it would be wrong.

Reply:
The existing documentation said that was undesirable, and that prefix iterators should only return values within the prefix, not a key exactly matching the prefix.

It could be that @jsimnz changed his mind after adding the documentation, and changed the badger code but not the memory store.

You can see the original documentation on (old) line 46-50 of `memory/iter.go`.

Reply:
This is a property that comes from the original `go-datastore/query` package we use. It uses this "strict prefix" semantics, so my intent was to keep those semantics the same for the entire CoreKV prefix iterator design.

Reply:
Discussed in standup, the team agreed that exact matches to the prefix should be returned.  I have updated/inversed https://github.com/sourcenetwork/corekv/issues/27 and will tackle this before integrating coreKV into Defra.